README.md
# hue-jazz
Assorted scripts and tricks for rooted Philips Hue hubs (and some related IOT). 

### Recent Updates
The latest version of the Hue firmware breaks the older Hue-Jazz remote control handling. Or maybe it's the remote controls' firmware. In any case, messages for
every button are sent each time you press one. That fooled the monitoring script into thinking you're pressing every button every time, causing the non-Hue lights to go berzerk.

I've updated to monitoring a different MQTT topic that can tell the difference again. If you use Hue-Jazz (and I don't know if anyone actually does--would love to know), then you'll need to update the script... or avoid the firmware update(s). The new version first assembles the entire list of dimmers from the appropriate MQTT topic and then uses them to map the individual button keys. This is unavoidable as the "publish" topic that this code relies upon does not say which button of a remote it's on except by its UUID (each button has a unique UUID). Copying all those UUIDs into the config seems like a pain so instead I deduced how they're mapped.

Note that Matter isn't going to make Hue remote controls work for controlling other devices. There still needs to be some handler in the Hub to translate the
remote requests into outside requests. However I have thought of a few other ways of doing this on rooted hubs which might be easier to maintain, such as forwarding the MQTT messages off the hub completely to a different device (freeing us from the tyranny of [Micro]python). There might even be hope for non-rooted hubs... I will look into all that when I'm more alert.

Long story short, your old settings will work but the rest of the code must be updated. I really need to externalize the config... my apologies.

### Prerequisites
The stuff here is designed to operate on a rooted Hue hub. You must first
be able to SSH to the Hub. 

* The best instructions for rooting the Hub are located here: 
[Philips Hue 2.1 enabling wifi](https://blog.andreibanaru.ro/2018/03/27/philips-hue-2-1-enabling-wifi/). You can stop after you've got SSH working. (You don't need a wifi adapter for my scripts, and with the latest Hue firmware, the wi-fi steps won't work anymore until someone can compile a working device driver; I'm trying.) Rooting the Hue Hub is fairly simple but cannot be performed without taking apart the unit and having the appropriate tools and a bit of basic hardware-hacking know-how. It would make for a good learning project if you have a spare Hue hub--and it is much easier than wiring new light switches. Prepare to say "eureka!" if you've never done this sort of thing before. It may become addictive. (By the way, I didn't use a soldering iron, I just stuck the serial connection wires in and made sure not to jiggle anything once I got the console and shorted the flash chip as instructed.) The unit seems fairly resilient but please do not blame me if you brick it.

* Micropython is required, but it's already on the Hue, at least in the current firmware (version 1946157000). You don't really need much coding skill to work with these scripts. 

* For the bulb_sync script you also need [rgbxy](https://github.com/benknight/hue-python-rgb-converter/). Whereever you put this script, you can just create a subdir called rgbxy and put that repo's \_\_init\_\_.py in it. 

* You will also probably need curl or wget on your Hue. I don't remember whether either one is on the Hue already, but they can be added via opkg (see [this article](https://medium.com/@rxseger/enabling-the-hidden-wi-fi-radio-on-the-philips-hue-bridge-2-0-42949f0154e1) for instructions on getting opkg running). 

* You can scp the script(s) to the hub, or just open up your text editor and paste it straight in from the clipboard.

* You will want to customize the settings toward the top of the script, obviously, then test with the simple: ```/usr/bin/micropython name_of_script.py```

* After you've tested your customizations, you can edit ```/etc/rc.local``` and add it to startup above the ```exit 0``` line:

* When it is all working and you want the Hue to automatically start your script, assuming you used bulb_sync.py and put in /root, just use:

```/usr/bin/micropython /root/bulb_sync.py &```

### What's here:

* **bulb_sync.py**: control a non-Hue bulb by updating another bulb with the conversion from its XY+brightness color to the HSV/HSB values needed in OpenHab (and probably other HA tools). This script monitors the queue for the change messages from a specific bulb, which means you can have multiple non-Hue lights do the same thing as any Hue bulb (triggered by the app or sensor), and when you use the dimmer the non-Hue lights will follow the same instructions with minimal latency. There's also optional code for a pointless color shift on the non-Hue bulb(s), i.e. silly lighting effects.

* **hue_mqtt_listener.py**: listen for specific sensor (dimmer) events and act on them. Can be combined with bulb_sync but doesn't require color conversion. This script takes a map of button commands and executes whatever arbitrary request you have in the map, most likely a curl command.

* Personally I have both scripts merged together for my purposes, but you can also just run both simultaneously if you need them.

### Background
My work began with the Hue dimmer switch and Openhab rules files. Here is my old, [very obsolete article](https://www.benchadwick.com/2018/12/using-a-hue-dimmer-switch-on-non-hue-devices-with-openhab2/) on how to make the switches work with OpenHab via the existing addon, but note that it has far higher latency than what you get through this hacked-Hue method.

My apartment doesn't have enough light switches, so IOT lights and wireless dimmers have given me a great deal of flexibility. The Hue dimmer is the best and cheapest solution for the switch problem, given that I cannot easily carve into the walls to rewire the place (and shouldn't have to, with IOT lights). 

The Hue Dimmer (especially the first version) is designed to adhere to walls (with tape or built-in magnets). I have not been able to find any affordable alternative that is as good for my needs, and Hue bulbs are expensive, so I've found ways to adapt the switch to control non-Hue lights through OpenHAB. 

The drawback of the OpenHAB rules-based approach described in my older article is that it must poll the Hue API for the state of the dimmers, which is not 100% reliable, since (for example) repeated button presses may be skipped, and which introduces additional latency (to what is already a flaky enough system due to wi-fi dropouts, a problem with the cheaper bulbs themselves).

The scripts here run on a rooted Hub itself and listen for the original signals coming over MQTT, which is as close as we can get to them. This means the dimmer can be a direct push trigger to whichever home automation software is in use (OpenHAB in my case; depending on the bulb and your network config, you could also send a request to its own API).

Dear Philips, love the products. Please never lock them down as a result of the six or seven people who will bother rooting the thing.
 
### Work in progress:
* improvements to the above scripts. Ideally they should load configuration from a separate file so no changes would be necessary to the main script. 
* attempts to insert a "virtual" light bulb or sensor that would be treated as real by the Hue app and API.
* attempts to restore the ath9k (or other?) driver for activating the built-in wireless unit. I was able to build the drivers for the kernel, but they still do not spot the chip (unlike in earlier versions of the Hue SOC).
* virtual "scenes" - a way to convince the unit to use non-Hue bulbs as parts of scenes.
* user-suggested tweaks

### Philosophies behind this work (in case you care):
* I bought a thing. I own it. I should be able to do whatever I want with it (if it doesn't harm anyone else)... Root access should be considered a right.
* The recent model of having in-home units make API calls to distant servers to control things in the home is idiotic. (Guilty: all Tuya/SmartHome products and pretty much all the other discount companies.) There may be exceptions for people who need to control stuff when they aren't home, but that should be the exception, not the standard. The only time I want my devices reaching out is when there are firmware upgrades and even then I want to know exactly what is changing, before it kicks in. So I'm trying to lock down all my IOT devices to never make outside calls without me knowing. In the case of the cheap stuff, I recommend flashing with Tasmota wherever possible. As that is become increasingly un-possible, I'm looking into other approaches.
* For environmental reasons and simple matters of sanity, Hue really needs to stop bundling extraneous Hubs in with the multi-packs. However the extra Hubs were an unexpected benefit this time around-- I now have three Hubs to experiment with... one rooted and stuck on the old firmware, one rooted with the new firmware, and one I haven't even pulled out of the sleeve.


CHANGELOG.md
Version 0.00000002 - updates to skirt the button last state problem
Version 0.00000001 - initial stuff, working until Hue added Matter support and restored button last state



bulb_sync/bulb_sync.py
#!/usr/bin/env python3
# bulb_sync.py -- from http://github.com/20goto10/hue-jazz
#
# Monitors a bulb for changes. You can use this to (fairly) easily tie a single Hue bulb with n non-Hue bulbs, so that
# all of them change in tandem (triggered by a dimmer, the app, or whatever).
# Requires: rgbxy, 
#   available here: https://github.com/benknight/hue-python-rgb-converter/ courtesy of Benjamin Knight.
#   You can just create a subdirectory called rgbxy and put that one __init__.py file in it.
#
# This is designed to be run on a rooted Hue with micropython.
#
# This software uses the Carl Perkins License, (C) 1955:
# Do anything that you want to do, but don't you step on my blue suede shoes.
#

import json
import uasyncio as asyncio
from rgbxy import *
from mqtt import Mqtt
import sys
import zlib
import os
import math

VERSION = "0.00000001"

# SETTINGS:
# Settings that you may need to modify. 
# At some point I will have this load from a separate JSON file.

# debug options: have it print messages for all the stuff it ignores, or all the stuff that it actually matches.
# I ship this with debugging on because you need some info from the MQTT messages, but once it's working you'll want this off.
print_skipped_messages = True
print_matched_messages = True
print_commands = True

# The source bulb's name in text (as saved in the Hue). This arrives in the
# metadata for the message. You must set this correctly or you won't see 
# any output. It is easy to discover by setting the above debug options to True.
name_of_bulb_to_monitor = "Dining room hue bulb" # for example

# Specify the Hue bulb's color Gamut here, identified as GamutA, GamutB, or GamutC.
# This can be determined by inspecting the bulb's status messages (and 
# and is probably tied to the model). My bulbs all use Gamut C so I'm not
# sure.  It's probably not a major issue if you have this wrong, it just 
# means your synced lights' colors won't line up as perfectly.
converter = Converter(GamutC)

# target bulb root of command. You will want to do some tailoring here. This should work on Openhab when you set it correctly.
# Other HA apps use similar but different APIs.
target_bulbs = ["MyLight"] # I doubt your bulb is also called "DiningHand"... you can add additional bulbs to this list.
# You can also trigger a cycling effect, if you want it, by including the Hue bulb (in Openhab) with a color shift, and changing
# color_shift to True, below, and having a number_of_colors value different from the actual number of bulbs.

# Should be noted, these don't actually need to be bulbs, unless you are writing a curl request to the actual item itself. 
# Otherwise, you don't need to list every bulb as a target, unless you want to use the color shift effect. 
# Otherwise just use a group/room/whatever in your home automation software and address the request to that. 
openhab_url = "http://lightcontroller:8080/"
target_bulb_cmd_prefixes = []
for bulb in target_bulbs:
  target_bulb_cmd_prefixes.append("/usr/bin/curl " + openhab_url + "rest/items/" + bulb + " -H 'Content-Type: text/plain' --data-raw")

# Optional effect to have the lights at alternate color wheel points, for balance or psychedelia or whatever. Included for silly fun.
# As written, each bulb in the list will be one more step along the phase. 
color_shift = True
phase_shift_angle = 60 

# END SETTINGS

async def data_listener(topic, mqtt):
    async for msg in mqtt.subscribe(topic):
        handle_data_message(topic, msg)                          
                  
async def main():
    async with Mqtt("pymosquitto") as m:
        tasks = []                                     
        tasks.append(asyncio.create_task(data_listener("dt/clip/+/+",m)))
        await asyncio.gather(*tasks)


def rgb_to_hsv(r, g, b):
    maxc = max(r, g, b)
    minc = min(r, g, b)
    v = maxc
    if minc == maxc:
        return 0.0, 0.0, v
    s = (maxc-minc) / maxc
    rc = (maxc-r) / (maxc-minc)
    gc = (maxc-g) / (maxc-minc)
    bc = (maxc-b) / (maxc-minc)
    if r == maxc:
        h = bc-gc
    elif g == maxc:
        h = 2.0+rc-bc
    else:
        h = 4.0+gc-rc
    h = (h/6.0) % 1.0
    return h, s, v


def handle_data_message(topic, msg):
      item = json.loads(zlib.decompress(msg["payload"], -15).decode())
          
      cmd_list = []

      # your handler stuff goes here -- one would probably want it to be a bit more sophisticated, but this is just a POC
      if "metadata" in item and "name" in item["metadata"]:
        if item["metadata"]["name"] == name_of_bulb_to_monitor:
          if print_matched_messages:
            print(topic, json.dumps(item, indent=2)) # you may want to remove the "indent" option to save screen space

          if "color" in item:
            # get color info from the reported changes 
            colors = item["color"]["xy"]
            bri = item["dimming"]["brightness"]
            bri_s = str(int(bri))
            is_on = item["on"]["on"] # not a typo

            # there's probably a way to go from XY straight to HSB, 
            # but eh, we're in no rush
            r,g,b = converter.xy_to_rgb(colors["x"], colors["y"], bri)
            h,s,v = rgb_to_hsv(r,g,b)
            h = h * 360.0 
            s = s * 100.0

            for cmd_prefix in target_bulb_cmd_prefixes:
              if color_shift:
                h = (int (h + phase_shift_angle)) % 360  
              if not is_on:
                color_string = "OFF" # sending any colors will trigger OpenHAB to turn it on, but Hue treats color and on/off independently, so this forces it to recognize on/off
              else:
                color_string = str(int(h)) + ","  + str(int(s)) + "," + bri_s
              cmd = cmd_prefix + " " + color_string
              cmd_list.append(cmd)

      else:
        if print_skipped_messages:
          print(topic, json.dumps(item, indent=2)) # you may want to remove the "indent" option to save screen space

      # execute the commands 
      for cmd in cmd_list:
        if print_commands:
          print("* Issuing command: " + cmd)
        os.system(cmd)

      sys.stdout.flush()

asyncio.run(main())
asyncio.get_event_loop().run_forever()



old_scripts/hue_tie_light_to_other_light.sh
#!/bin/sh
# This simple script does not require a rooted Hue but it does require
# jsonfilter, which is *on* the Hub, and does not seem to be the same
# project that exists on Github. I think the right one is available via opkg 
# for other, similar SOC devices.
#
# This script polls OpenHab every 1 second for the status of one bulb and uses
# it to update another bulb. Consider it the first draft of the faster,
# better, rooted-Hue only micropython script available at the hue-jazz repo.
#
# Find more info at: github.com/20goto10/hue-jazz
#
# It is also probably possible to do this through OpenHab itself, so 
# basically, I don't think there is any real audience for this script.
# "jq" can do the same thing as "jsonfilter" if you do need this.

### Parameters to set yourself:

OPENHAB_URL=http://whatever:8080 # update this to your OpenHAB url
HUE_API_KEY=some_key # you get this key by registering a user through the API (i.e. after pressing the button on the Hue)
HUE_API_URL=http://localhost # the IP address of the Hue Hub, or "http://localhost" if you are running this on the Hue itself

# Hue API ID for the light being tracked
TRACKED_BULB_ID=/lights/5

# Openhab URL of the light being updated
UPDATED_BULB_URL=${OPENHAB_URL}/rest/items/MyNotHueBulb

### 

while true
  do
    # state is written to a temp file b/c I could not get  
    # jsonfilter to work correctly on the simple string (quote escape
    # issues not worth figuring out)
    tfile="/tmp/curled_insta_state"
    echo curl -s ${HUE_API_URL}/api/${HUE_API_KEY}/${TRACKED_BULB_ID}
    curl -s ${HUE_API_URL}/api/${HUE_API_KEY}/${TRACKED_BULB_ID} > $tfile

#    echo `cat $tfile`
    hue=$(jsonfilter -i $tfile -e '@["state"]["hue"]')
    sat=$(jsonfilter -i $tfile -e '@["state"]["sat"]')
    bri=$(jsonfilter -i $tfile -e '@["state"]["bri"]')

    newhue=$(expr $hue / 182)
    newsat=$(expr $sat \* 100 / 255)
    newbri=$(expr $bri \* 100 / 255)
    result="$newhue,$newsat,$newbri"
    # echo curl $UPDATED_BULB_URL -H 'Content-Type: text/plain' --data-raw $result 
    curl $UPDATED_BULB_URL -H 'Content-Type: text/plain' --data-raw $result 
    sleep 1
  done



hue_mqtt_monitor/hue_mqtt_monitor.py
#!/usr/bin/env python3
# hue_mqtt_monitor.py -- from http://github.com/20goto10/hue-jazz
#
# Trigger whatever requests you like (presumably to an alternate server) from a dimmer switch or other sensor.
# It is designed to run on a rooted Hue using micropython.
#
# This software uses the Carl Perkins License, (C) 1955:
# Do anything that you want to do, but don't you step on my blue suede shoes.

import json
import uasyncio as asyncio
from mqtt import Mqtt
import sys
import zlib
import os
import math

VERSION = "0.00000002"

# SETTINGS:
# Settings that you may need to modify. 
# At some point I will have this load from a separate JSON file.

# debug options: have it print messages for all the stuff it ignores, or all the stuff that it actually matches.
# I ship this with debugging on because you need some info from the MQTT messages, but once it's working you'll want this off.
print_everything = True
print_skipped_messages = False 
print_matched_messages = True
print_commands = True 

# This is set up for Openhab, but it could be anything you want that has a reachable API... or fire off some other command altogether.
openhab_url = "http://mylightcontroller:8080/"
curl_command = "/usr/bin/curl"

def curl_req_for(target, command):
  return curl_command + " " + openhab_url + "rest/items/" + target + " -H 'Content-Type: text/plain' --data-raw " + command

# Map of requests (in my format) to actions. The /sensors/id can be derived from the debug output. Each dimmer has four buttons, 1-4.
# The types of presses are initial_press, short_release, long_release, repeat. 'repeat' seems to show up for long_release only and I'm not sure what it really means.
# The key is the request fields for the sensor ID, button press type, and button # on the dimmer, delineated by colons.
action_map = {
              # example: map top and bottom to on/off for some group, middle buttons on/off for some light
              "/sensors/49:short_release:1": curl_req_for("LivingRoom", "ON"),
              "/sensors/49:short_release:2": curl_req_for("Light1", "ON"),
              "/sensors/49:short_release:3": curl_req_for("Light1", "OFF"),
              "/sensors/49:short_release:4": curl_req_for("LivingRoom", "OFF"),

              # example: have different effects on long-press on a different dimmre
              "/sensors/50:short_release:1": curl_req_for("ColLivRoomLights", "ON"),
              "/sensors/50:short_release:2": curl_req_for("CouchLight", "ON"),
              "/sensors/50:long_release:2":  curl_req_for("StageLight", "ON"),
              "/sensors/50:short_release:3": curl_req_for("CouchLight", "OFF"),
              "/sensors/50:long_release:3":  curl_req_for("StageLight", "OFF"),
              "/sensors/50:short_release:4": curl_req_for("ColLivRoomLights", "OFF"),

              # example: an arbitrary curl request to some outside API (i.e. control some Tuya/SmartHome/Govee/whatever unsupported by OpenHab)
              "/sensors/14:short_release:1": "curl --insecure --location --request PUT 'https://some-random-tuya-api-or-whatever' --header 'My-API-Key: blahblah' -- developer-api.doyoureallytrustthissystem.com/v1/devices/control' 'Content-Type: application/json' --data-raw '{\"device\": \"AA:BB:CC:DD:EE:FF:11:22\",\"model\": \"Banana\",\"cmd\": {\"name\": \"turn\", \"value\": \"on\"}}'",
              "/sensors/14:short_release:4": "curl --insecure --location --request PUT 'https://some-random-tuya-api-or-whatever' --header 'My-API-Key: blahblah' -- developer-api.doyoureallytrustthissystem.com/v1/devices/control' 'Content-Type: application/json' --data-raw '{\"device\": \"AA:BB:CC:DD:EE:FF:11:22\",\"model\": \"Banana\",\"cmd\": {\"name\": \"turn\", \"value\": \"off\"}}'",
             }
# END SETTINGS

device_map = {}

async def device_listener(mqtt):
    topic = "dt/clip/device/+"
    async for msg in mqtt.subscribe(topic):
        handle_sensor(topic, msg)

async def data_listener(topic, mqtt):
    async for msg in mqtt.subscribe(topic):
        handle_data_message(topic, msg)
                  
async def main():
    async with Mqtt("pymosquitto") as m:
        tasks = []                                     
        tasks.append(asyncio.create_task(device_listener(m)))
        tasks.append(asyncio.create_task(data_listener("cmd/clip/event/publish",m)))
        await asyncio.gather(*tasks)


def handle_sensor(topic, msg):
   try:
      item = json.loads(zlib.decompress(msg["payload"], -15).decode())
      if print_everything:
        print("device found", topic, item)
      if item["product_data"]["product_name"] == "Hue dimmer switch":
        if "services" in item:
          bulbs = item["services"]
          counter = 1 
          for service in bulbs:
            if service["rtype"] == "button":
              if print_everything:
                print("Adding", service["rtype"], service["rid"])
              device_map[service["rid"]] = str(counter)
              counter = counter + 1
   except Exception as e: 
      print("handle_sensor error:", str(e))


def handle_data_message(topic, msg):
   try:
      item = json.loads(msg["payload"].decode())

      if print_everything:
        print(item)
          
      cmd_list = []


      # your handler stuff goes here -- one would probably want it to be a bit more sophisticated, but this is just a POC
      # sensor button handling
      if "data" in item and "button" in item["data"][0]["data"][0]:
        id_v1 = item["data"][0]["data"][0]["id_v1"]
        last_event = item["data"][0]["data"][0]["button"]["last_event"]
        control_id = device_map[item["data"][0]["data"][0]["id"]]
        request = id_v1 + ":" + last_event + ":" + control_id

        if request in action_map:
          cmd_list.append(action_map[request])

        if print_matched_messages:
          print(request) # use this to easily map as-yet unmapped buttons
        
      else:
        if print_skipped_messages:
          print(topic, json.dumps(item, indent=2)) # you may want to remove the "indent" option to save screen space

      # execute the commands 
      for cmd in cmd_list:
        if print_commands:
          print("* Issuing command: %s" % cmd)
        try:
          os.system(cmd)
        except:
          print("* Command failed %s" % cmd)
      sys.stdout.flush()
   except Exception as e: 
      print("Execution failure", str(e))

asyncio.run(main())
asyncio.get_event_loop().run_forever()


